File: Day027_2025/Q027_Leetcode.java
import java.util.*;

class Q027_Leetcode {
    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        // Create adjacency list and indegree array
        Map<Integer, List<Integer>> adjList = new HashMap<>();
        int[] indegree = new int[numCourses];
        for (int[] edge : prerequisites) {
            adjList.computeIfAbsent(edge[0], _ -> new ArrayList<>()).add(edge[1]);
            indegree[edge[1]]++;
        }

        // Initialize queue for topological sort
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.add(i);
            }
        }

        // Map to store prerequisites for each course
        Map<Integer, Set<Integer>> prereqMap = new HashMap<>();

        // Topological sort to build prerequisite map
        while (!queue.isEmpty()) {
            int course = queue.poll();
            for (int nextCourse : adjList.getOrDefault(course, new ArrayList<>())) {
                prereqMap.computeIfAbsent(nextCourse, _ -> new HashSet<>()).add(course);
                prereqMap.get(nextCourse).addAll(prereqMap.getOrDefault(course, new HashSet<>()));
                if (--indegree[nextCourse] == 0) {
                    queue.add(nextCourse);
                }
            }
        }

        // Answer queries
        List<Boolean> result = new ArrayList<>();
        for (int[] query : queries) {
            result.add(prereqMap.getOrDefault(query[1], new HashSet<>()).contains(query[0]));
        }

        return result;
    }
}

File: Day027_2025/Q027_Leetcode.py
from typing import List
from collections import defaultdict, deque

class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        # Create adjacency list and indegree array
        adj_list = defaultdict(list)
        indegree = [0] * numCourses
        for u, v in prerequisites:
            adj_list[u].append(v)
            indegree[v] += 1

        # Initialize queue for topological sort
        queue = deque([i for i in range(numCourses) if indegree[i] == 0])

        # Map to store prerequisites for each course
        prereq_map = defaultdict(set)

        # Topological sort to build prerequisite map
        while queue:
            course = queue.popleft()
            for next_course in adj_list[course]:
                prereq_map[next_course].add(course)
                prereq_map[next_course].update(prereq_map[course])
                indegree[next_course] -= 1
                if indegree[next_course] == 0:
                    queue.append(next_course)

        # Answer queries
        return [query[0] in prereq_map[query[1]] for query in queries]
